File: FE-RESTAURANT/src/services/VoucherUsageHistoryService.ts
import apiClient from '../utils/apiClient';

export interface VoucherUsageHistory {
  id: number;
  voucherId: number;
  voucherName: string;
  voucherCode: string;
  orderId: number;
  customerPhone?: string;
  customerName?: string;
  originalAmount: number;
  discountAmount: number;
  finalAmount: number;
  usedAt: string;
  createdAt: string;
  updatedAt: string;
}

export interface VoucherUsageHistoryFilters {
  voucherCode?: string;
  customerPhone?: string;
  startDate?: string;
  endDate?: string;
}

export interface PaginatedResponse<T> {
  content: T[];
  totalElements: number;
  totalPages: number;
  size: number;
  number: number;
  first: boolean;
  last: boolean;
}

class VoucherUsageHistoryService {
  // Lấy tất cả lịch sử sử dụng voucher với phân trang
  async getAllVoucherUsageHistory(
    page: number = 0,
    size: number = 10,
    filters?: VoucherUsageHistoryFilters
  ): Promise<PaginatedResponse<VoucherUsageHistory>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        size: size.toString()
      });

      if (filters?.voucherCode) params.append('voucherCode', filters.voucherCode);
      if (filters?.customerPhone) params.append('customerPhone', filters.customerPhone);
      if (filters?.startDate) params.append('startDate', filters.startDate);
      if (filters?.endDate) params.append('endDate', filters.endDate);

      const response = await apiClient.get(`/voucher-usage-history?${params}`);
      console.log('Raw API Response:', response.data); // Debug log
      return response.data;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Lấy lịch sử sử dụng voucher theo voucher ID
  async getVoucherUsageHistory(voucherId: number): Promise<VoucherUsageHistory[]> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/voucher/${voucherId}`);
      return response.data;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Lấy lịch sử sử dụng voucher theo order ID
  async getVoucherUsageByOrderId(orderId: number): Promise<VoucherUsageHistory | null> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/order/${orderId}`);
      return response.data;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Lấy lịch sử sử dụng voucher theo mã voucher
  async getVoucherUsageByCode(voucherCode: string): Promise<VoucherUsageHistory[]> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/code/${voucherCode}`);
      return response.data;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Lấy lịch sử sử dụng voucher theo số điện thoại khách hàng
  async getVoucherUsageByCustomerPhone(customerPhone: string): Promise<VoucherUsageHistory[]> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/customer/${customerPhone}`);
      return response.data;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Lấy lịch sử sử dụng voucher trong khoảng thời gian
  async getVoucherUsageByDateRange(startDate: string, endDate: string): Promise<VoucherUsageHistory[]> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/date-range?startDate=${startDate}&endDate=${endDate}`);
      return response.data;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Lấy voucher được sử dụng nhiều nhất
  async getMostUsedVouchers(limit: number = 10): Promise<any[]> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/most-used-vouchers?limit=${limit}`);
      return response.data;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Lấy khách hàng sử dụng voucher nhiều nhất
  async getMostActiveCustomers(limit: number = 10): Promise<any[]> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/most-active-customers?limit=${limit}`);
      return response.data;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Đếm số lần sử dụng voucher
  async getVoucherUsageCount(voucherId: number): Promise<number> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/voucher/${voucherId}/count`);
      return response.data.usageCount;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }

  // Đếm số lần sử dụng voucher theo mã
  async getVoucherUsageCountByCode(voucherCode: string): Promise<number> {
    try {
      const response = await apiClient.get(`/voucher-usage-history/code/${voucherCode}/count`);
      return response.data.usageCount;
    } catch (error: any) {
      console.error('❌ Voucher Usage History API error:', error.response?.data || error.message);
      throw error;
    }
  }
}

export { VoucherUsageHistoryService };
export const voucherUsageHistoryService = new VoucherUsageHistoryService();

// Export individual methods for easier use
export const getAllVoucherUsageHistory = voucherUsageHistoryService.getAllVoucherUsageHistory.bind(voucherUsageHistoryService);
export const getVoucherUsageHistory = voucherUsageHistoryService.getVoucherUsageHistory.bind(voucherUsageHistoryService);
export const getVoucherUsageByOrderId = voucherUsageHistoryService.getVoucherUsageByOrderId.bind(voucherUsageHistoryService);
export const getVoucherUsageByCode = voucherUsageHistoryService.getVoucherUsageByCode.bind(voucherUsageHistoryService);
export const getVoucherUsageByCustomerPhone = voucherUsageHistoryService.getVoucherUsageByCustomerPhone.bind(voucherUsageHistoryService);
export const getVoucherUsageByDateRange = voucherUsageHistoryService.getVoucherUsageByDateRange.bind(voucherUsageHistoryService);
export const getMostUsedVouchers = voucherUsageHistoryService.getMostUsedVouchers.bind(voucherUsageHistoryService);
export const getMostActiveCustomers = voucherUsageHistoryService.getMostActiveCustomers.bind(voucherUsageHistoryService);
export const getVoucherUsageCount = voucherUsageHistoryService.getVoucherUsageCount.bind(voucherUsageHistoryService);
export const getVoucherUsageCountByCode = voucherUsageHistoryService.getVoucherUsageCountByCode.bind(voucherUsageHistoryService);

File: RESTAURANT/src/main/java/com/poly/restaurant/services/VoucherUsageHistoryService.java
package com.poly.restaurant.services;

import com.poly.restaurant.dtos.VoucherUsageHistoryRequestDTO;
import com.poly.restaurant.dtos.VoucherUsageHistoryResponseDTO;
import com.poly.restaurant.entities.VoucherUsageHistoryEntity;
import com.poly.restaurant.entities.DiscountEntity;
import com.poly.restaurant.entities.OrderEntity;
import com.poly.restaurant.exceptions.ResourceNotFoundException;
import com.poly.restaurant.mappers.VoucherUsageHistoryMapper;
import com.poly.restaurant.repositories.VoucherUsageHistoryRepository;
import com.poly.restaurant.repositories.DiscountRepository;
import com.poly.restaurant.repositories.OrderRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class VoucherUsageHistoryService {

    private static final Logger logger = LoggerFactory.getLogger(VoucherUsageHistoryService.class);
    
    private final VoucherUsageHistoryRepository voucherUsageHistoryRepository;
    private final DiscountRepository discountRepository;
    private final OrderRepository orderRepository;

    /**
     * Lưu lịch sử sử dụng voucher
     */
    @Transactional
    public VoucherUsageHistoryResponseDTO saveVoucherUsage(VoucherUsageHistoryRequestDTO request) {
        logger.info("Saving voucher usage history for voucher: {}, order: {}", 
                   request.getVoucherId(), request.getOrderId());
        
        // Lấy voucher và order
        DiscountEntity voucher = discountRepository.findById(request.getVoucherId())
                .orElseThrow(() -> new ResourceNotFoundException("Voucher not found with id: " + request.getVoucherId()));
        
        OrderEntity order = orderRepository.findById(request.getOrderId())
                .orElseThrow(() -> new ResourceNotFoundException("Order not found with id: " + request.getOrderId()));
        
        // Tạo entity
        VoucherUsageHistoryEntity entity = VoucherUsageHistoryMapper.toEntity(request, voucher, order);
        
        // Lưu vào database
        VoucherUsageHistoryEntity savedEntity = voucherUsageHistoryRepository.save(entity);
        
        logger.info("Voucher usage history saved with ID: {}", savedEntity.getId());
        
        return VoucherUsageHistoryMapper.toResponseDto(savedEntity);
    }

    /**
     * Lấy lịch sử sử dụng voucher theo voucher ID
     */
    @Transactional(readOnly = true)
    public List<VoucherUsageHistoryResponseDTO> getVoucherUsageHistory(Long voucherId) {
        logger.info("Getting voucher usage history for voucher: {}", voucherId);
        
        List<VoucherUsageHistoryEntity> entities = voucherUsageHistoryRepository.findByVoucherIdOrderByUsedAtDesc(voucherId);
        
        return entities.stream()
                .map(VoucherUsageHistoryMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    /**
     * Lấy lịch sử sử dụng voucher theo order ID
     */
    @Transactional(readOnly = true)
    public Optional<VoucherUsageHistoryResponseDTO> getVoucherUsageByOrderId(Long orderId) {
        logger.info("Getting voucher usage history for order: {}", orderId);
        
        Optional<VoucherUsageHistoryEntity> entity = voucherUsageHistoryRepository.findByOrderId(orderId);
        
        return entity.map(VoucherUsageHistoryMapper::toResponseDto);
    }

    /**
     * Lấy lịch sử sử dụng voucher theo mã voucher
     */
    @Transactional(readOnly = true)
    public List<VoucherUsageHistoryResponseDTO> getVoucherUsageByCode(String voucherCode) {
        logger.info("Getting voucher usage history for voucher code: {}", voucherCode);
        
        List<VoucherUsageHistoryEntity> entities = voucherUsageHistoryRepository.findByVoucherCodeOrderByUsedAtDesc(voucherCode);
        
        return entities.stream()
                .map(VoucherUsageHistoryMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    /**
     * Lấy lịch sử sử dụng voucher theo số điện thoại khách hàng
     */
    @Transactional(readOnly = true)
    public List<VoucherUsageHistoryResponseDTO> getVoucherUsageByCustomerPhone(String customerPhone) {
        logger.info("Getting voucher usage history for customer phone: {}", customerPhone);
        
        List<VoucherUsageHistoryEntity> entities = voucherUsageHistoryRepository.findByCustomerPhoneOrderByUsedAtDesc(customerPhone);
        
        return entities.stream()
                .map(VoucherUsageHistoryMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    /**
     * Lấy lịch sử sử dụng voucher trong khoảng thời gian
     */
    @Transactional(readOnly = true)
    public List<VoucherUsageHistoryResponseDTO> getVoucherUsageByDateRange(LocalDateTime startDate, LocalDateTime endDate) {
        logger.info("Getting voucher usage history from {} to {}", startDate, endDate);
        
        List<VoucherUsageHistoryEntity> entities = voucherUsageHistoryRepository.findByUsedAtBetween(startDate, endDate);
        
        return entities.stream()
                .map(VoucherUsageHistoryMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    /**
     * Lấy lịch sử sử dụng voucher với phân trang
     */
    @Transactional(readOnly = true)
    public Page<VoucherUsageHistoryResponseDTO> getVoucherUsageHistoryPaginated(Long voucherId, int page, int size) {
        logger.info("Getting paginated voucher usage history for voucher: {}, page: {}, size: {}", 
                   voucherId, page, size);
        
        Pageable pageable = PageRequest.of(page, size);
        Page<VoucherUsageHistoryEntity> entityPage = voucherUsageHistoryRepository.findByVoucherIdOrderByUsedAtDesc(voucherId, pageable);
        
        return entityPage.map(VoucherUsageHistoryMapper::toResponseDto);
    }

    /**
     * Lấy tất cả lịch sử sử dụng voucher với phân trang (cho trang quản lý)
     */
    @Transactional(readOnly = true)
    public Page<VoucherUsageHistoryResponseDTO> getAllVoucherUsageHistoryPaginated(int page, int size) {
        logger.info("Getting all voucher usage history with pagination - page: {}, size: {}", page, size);
        
        Pageable pageable = PageRequest.of(page, size);
        Page<VoucherUsageHistoryEntity> entityPage = voucherUsageHistoryRepository.findAll(pageable);
        
        return entityPage.map(VoucherUsageHistoryMapper::toResponseDto);
    }

    /**
     * Lấy lịch sử sử dụng voucher với filter và phân trang
     */
    @Transactional(readOnly = true)
    public Page<VoucherUsageHistoryResponseDTO> getVoucherUsageHistoryWithFilters(
            String voucherCode, String customerPhone, LocalDateTime startDate, LocalDateTime endDate,
            int page, int size) {
        logger.info("Getting voucher usage history with filters - voucherCode: {}, customerPhone: {}, startDate: {}, endDate: {}, page: {}, size: {}", 
                   voucherCode, customerPhone, startDate, endDate, page, size);
        
        Pageable pageable = PageRequest.of(page, size);
        Page<VoucherUsageHistoryEntity> entityPage = voucherUsageHistoryRepository.findByFilters(
                voucherCode, customerPhone, startDate, endDate, pageable);
        
        return entityPage.map(VoucherUsageHistoryMapper::toResponseDto);
    }

    /**
     * Đếm số lần sử dụng voucher
     */
    @Transactional(readOnly = true)
    public long getVoucherUsageCount(Long voucherId) {
        return voucherUsageHistoryRepository.countByVoucherId(voucherId);
    }

    /**
     * Đếm số lần sử dụng voucher theo mã
     */
    @Transactional(readOnly = true)
    public long getVoucherUsageCountByCode(String voucherCode) {
        return voucherUsageHistoryRepository.countByVoucherCode(voucherCode);
    }

    /**
     * Lấy voucher được sử dụng nhiều nhất
     */
    @Transactional(readOnly = true)
    public List<Object[]> getMostUsedVouchers(int limit) {
        logger.info("Getting most used vouchers with limit: {}", limit);
        
        Pageable pageable = PageRequest.of(0, limit);
        return voucherUsageHistoryRepository.findMostUsedVouchers(pageable);
    }

    /**
     * Lấy khách hàng sử dụng voucher nhiều nhất
     */
    @Transactional(readOnly = true)
    public List<Object[]> getMostActiveCustomers(int limit) {
        logger.info("Getting most active customers with limit: {}", limit);
        
        Pageable pageable = PageRequest.of(0, limit);
        return voucherUsageHistoryRepository.findMostActiveCustomers(pageable);
    }
}
